elfi/methods/methods.py

```diff
@@ -473,13 +466,169 @@ def _merge_batch(self, batch):
+        """Updates the objective n_batches if applicable"""
+        if not self.objective.get('threshold'): return
+
+        s = self.state
+        t, n_samples = [self.objective.get(k) for k in ('threshold', 'n_samples')]
+
+        # noinspection PyTypeChecker
+        n_acceptable = np.sum(s['samples'][self.discrepancy] <= t) if s['samples'] else 0
+        if n_acceptable == 0: return
+
+        accept_rate_t = n_acceptable / s['n_sim']
+        # Add some margin to estimated batches_total. One could use confidence bounds here
+        margin = .2 * self.batch_size * int(n_acceptable < n_samples)
+        n_batches = (n_samples / accept_rate_t + margin) / self.batch_size
+
+        self.objective['n_batches'] = ceil(n_batches)
+        logger.debug('Estimated objective n_batches=%d' % self.objective['n_batches'])
+
+
+class SMC(Sampler):
+    def __init__(self, model, discrepancy=None, outputs=None, **kwargs):
+        model, self.discrepancy = self._resolve_model(model, discrepancy)
+        outputs = self._ensure_outputs(outputs, model.parameters + [self.discrepancy])
+        model, added_nodes = self._augment_model(model)
+
+        super(SMC, self).__init__(model, outputs + added_nodes, **kwargs)
+
+        self.state['round'] = 0
+        self._populations = []
+        self._rejection = None
+
+    def init_inference(self, n_samples, thresholds):
+        self.objective.update(dict(n_samples=n_samples,
+                                   n_batches=self.max_parallel_batches,
+                                   round=len(thresholds) - 1,
+                                   thresholds=thresholds))
+        self._new_round()
+
+    def extract_result(self):
+        # TODO: make a better result object
+        pop = self._extract_population()
+        result = self.state.copy()
+        result['populations'] = self._populations.copy()
+        result['populations'].append(pop)
+        result['samples'] = pop['samples']
+        return result
+
+    def update(self, batch, batch_index):
+        self._rejection.update(batch, batch_index)
+
+        if self._rejection.finished:
+            self.batches.reset(self.state['n_batches'] + 1)
+            if self.state['round'] < self.objective['round']:
+                self._populations.append(self._extract_population())
+                self.state['round'] += 1
+                self._new_round()
+
+        self._update_state()
+        self._update_objective()
+
+    def prepare_new_batch(self, batch_index):
+        # Use the actual prior
+        if self.state['round'] == 0:
+            return
+
+        # Sample from the proposal
+        params = GMDistribution.rvs(*self._gm_params, size=self.batch_size)
+        # TODO: support vector parameter nodes
+        batch = {p:params[:,i] for i, p in enumerate(self.parameters)}
+        return batch
+
+    def _new_round(self):
+        dashes = '-'*16
+        logger.info('%s Starting round %d %s' % (dashes, self.state['round'], dashes))
```

Comment:
This is great. Maybe we could implement something similar in a more general context for remaining batches.